### PR TITLE
 feat(connector): [Stripe] implement Multibanco Bank Transfer for stripe 

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Multibanco is a portuguese interbank network. It is the largest interbank network in Portugal.

Steps followed for Multibanco transfers:

1.After getting payments/confirm call for Multibanco bank transfer, stripe's source endpoint is called. We get the reference and entity numbers. This response from stripe's endpoint is shown in the Hyperswitch's confirm call response.
```
"multibanco": {
                "reference": "999999999",
                "entity": "12345"
            }
```

2.Front-end needs to instruct the customer to transfer funds from their bank to the above reference number and entity number
3. Once the customer transfers money from their account and the transfer is settled, the above source becomes chargeable and Stripe sends the source.chargeable and source.transaction.created webhook events.
4. After getting the above mentioned webhook events, we call customer endpoint and get the id and consume it in charges call.
5. The charge immediately succeeds but in case of some unexpected error, charge.succeeded webhook event from Stripe is received to confirm the payment success.
6. After the successful charge, webhook is triggered to merchant.

Curl For Testing

```
curl --location '<base_url>/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: <api_key>' \
--data-raw '{
  "amount": 6540,
  "currency": "EUR",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "amount_to_capture": 6540,
  "customer_id": "StripeCustomer",
  "email": "guest@example.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+1",
  "description": "Its my first payment request",
  "authentication_type": "no_three_ds",
  "return_url": "https://google.com",
  "payment_method": "bank_transfer",
  "payment_method_type": "multibanco",
  "payment_method_data": {
    "bank_transfer": {
        "multibanco_bank_transfer" : {
            "billing_details" : {
                "email" : "someone@something.com"
            }
        }
    }
  },
  "billing": {
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US",
      "first_name": "joseph",
      "last_name": "Doe"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+91"
    }
  },
  "shipping": {
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US",
      "first_name": "joseph",
      "last_name": "Doe"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+91"
    }
  },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "metadata": {
    "udf1": "value1",
    "new_customer": "true",
    "login_date": "2019-09-10T10:11:12Z"
  }
}'

```
### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Bank Transfers are a popular way of transmitting money between different bank accounts and they are popular in US, EU and few Asian and LATAM countries. They are primarily used by businesses for accepting large payments from other businesses. Bank transfers are also used by consumers in certain countries as a preferred method for transferring money to others and while transacting online.
So, Multibanco bank transfer will be added as a bank transfer feature

## How did you test it?
Tested it through Postman
Screen Shot 2023-06-13 at 8 17 18 AM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/stripe/transformers.rs [hunk 1]
```diff
@@ -2113,23 +2162,71 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
+                            ),
+                            return_url: Some(
+                                item.return_url
+                                    .clone()
+                                    .get_required_value("return_url")
+                                    .change_context(
+                                        errors::ConnectorError::MissingRequiredField {
+                                            field_name: "return_url",
+                                        },
+                                    )?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2207]:
```suggestion
                                item.get_return_url()?,
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 2]
```diff
@@ -2113,23 +2162,71 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
-                })?
-                .to_string(),
-        })
+        let currency = item
+            .request
+            .currency
+            .get_required_value("currency")
+            .change_context(errors::ConnectorError::MissingRequiredField {
+                field_name: "currency",
+            })?
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2174]:
```suggestion
        let currency = item.request.get_currency()?
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 3]
```diff
@@ -2113,23 +2162,71 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
+            Some(payments::PaymentMethodData::BankTransfer(bank_transfer_data)) => {
+                match **bank_transfer_data {
+                    payments::BankTransferData::MultibancoBankTransfer { .. } => Ok(
+                        Self::MultibancoBankTansfer(MultibancoCreditTransferSourceRequest {
+                            transfer_type: StripePaymentMethodType::Multibanco,
+                            currency,
+                            payment_method_data: MultibancoTransferData {
+                                email: connector::utils::PaymentsPreProcessingData::get_email(
+                                    &item.request,
+                                )?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2187]:
```suggestion
                                email: item.request.get_email()?,
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 4]
```diff
@@ -2113,23 +2162,71 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
+                            },
+                            amount: Some(
+                                item.request
+                                    .amount
+                                    .get_required_value("amount")
+                                    .change_context(
+                                        errors::ConnectorError::MissingRequiredField {
+                                            field_name: "amount",
+                                        },
+                                    )?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2197]:
```suggestion
                                item.request.get_amount()?,
```

File: crates/router/src/connector/stripe/transformers.rs [hunk 5]
```diff
@@ -2113,23 +2162,71 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
+                            ),
+                        }),
+                    ),
+                    payments::BankTransferData::AchBankTransfer { .. } => {
+                        Ok(Self::AchBankTansfer(AchCreditTransferSourceRequest {
+                            transfer_type: StripePaymentMethodType::AchCreditTransfer,
+                            payment_method_data: AchTransferData {
+                                email: connector::utils::PaymentsPreProcessingData::get_email(
+                                    &item.request,
+                                )?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:2217]:
```suggestion
                                email: item.request.get_email()?,
```

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Review by srujanchikke — DISMISSED** [review]

File: crates/router/src/connector/stripe.rs [hunk 6]
```diff
@@ -130,12 +130,14 @@ impl
         &self,
         req: &types::PaymentsPreProcessingRouterData,
     ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
-        let req = stripe::StripeAchSourceRequest::try_from(req)?;
+        let req = stripe::StripeCreditTransferSourceRequest::try_from(req)?;
+        router_env::logger::info!(connector_request=?req);
```
Comment (SanchithHegde) [inline crates/router/src/connector/stripe.rs:134]: Is logging of the connector request intentional?

File: crates/router/src/connector/stripe/transformers.rs [hunk 7]
```diff
@@ -2125,23 +2171,42 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
-                .currency
-                .get_required_value("currency")
-                .change_context(errors::ConnectorError::MissingRequiredField {
-                    field_name: "currency",
-                })?
-                .to_string(),
-        })
+        let currency = item.request.get_currency()?.to_string();
+
+        match &item.request.payment_method_data {
```
Comment (SanchithHegde) [inline crates/router/src/connector/stripe/transformers.rs:2179]: `item` is already a reference, you don't need the `&` again.

File: crates/router/src/connector/stripe.rs (hunk 6 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/stripe.rs:134]: No. I will remove it

File: crates/router/src/connector/stripe/transformers.rs (hunk 7 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/stripe/transformers.rs:2179]: Removing `&` throws an error

**Review by srujanchikke — DISMISSED** [review]

**Review by srujanchikke — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/connector/stripe/transformers.rs [hunk 8]
```diff
@@ -2127,23 +2173,42 @@ impl TryFrom<&types::PaymentsCaptureRouterData> for CaptureRequest {
...
-            currency: item
-                .request
-                .currency
-                .get_required_value("currency")
-                .change_context(errors::ConnectorError::MissingRequiredField {
-                    field_name: "currency",
-                })?
-                .to_string(),
-        })
+        let currency = item.request.get_currency()?.to_string();
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stripe/transformers.rs:2179]: Lets change this to enum instead of string

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — DISMISSED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Pull request merged**